avocado/utils/asset.py

```diff
@@ -71,21 +71,178 @@ def __init__(self, name, asset_hash, algorithm, locations, cache_dirs,
         :param expire: time in seconds for the asset to expire
         :param metadata: metadata which will be saved inside metadata file
         """
-        self.name = name
+        parsed_url = urlparse.urlparse(name)
+        # Asset file name
+        self.basename = os.path.basename(parsed_url.path)
         self.asset_hash = asset_hash
-        if algorithm is None:
-            self.algorithm = DEFAULT_HASH_ALGORITHM
-        else:
+
+        if algorithm:
             self.algorithm = algorithm
+        else:
+            self.algorithm = DEFAULT_HASH_ALGORITHM
 
+        self.locations = []
+        # If `name` is an url, move it to `locations`
+        if parsed_url.scheme:
+            self.locations.append(os.path.dirname(parsed_url.geturl()))
+        # Let's be conservative and check `locations` type
         if isinstance(locations, str):
-            self.locations = [locations]
+            # remove asset name from location
+            if self.basename in locations:
+                parsed_locations = urlparse.urlparse(locations)
+                self.locations.append(
+                    os.path.dirname(parsed_locations.geturl()))
+            else:
+                self.locations.append(locations)
+        elif isinstance(locations, list):
+            # remove asset name from location
+            for location in locations:
+                if self.basename in location:
+                    parsed_location = urlparse.urlparse(location)
+                    self.locations.append(
+                        os.path.dirname(parsed_location.geturl()))
+                else:
+                    self.locations.append(location)
+
+        # Let's make the user's life easy and handle the cache
+        if cache_dirs:
+            self.cache_dirs = cache_dirs
         else:
-            self.locations = locations
-        self.cache_dirs = cache_dirs
+            # Break early if we don't have the cache_dirs. This is necessary
```

Comment:
If you don't want to break the API, why did you make cache_dirs optional?

Reply:
My willing was to change the class signature to `Asset(name, cache_dirs, asset_hash=None, algorithm=None, locations=None, expire=None, metadata=None)` as the required parameters with this refactor are `name` and `cache_dirs`, but this breaks the tests already implemented using the original class signature.

As I cannot do `cache_dirs` required with the order of the parameters in the original class signature, letting `asset_hash`, `algorithm`, and  `locations` as optional parameters, the solution I found was to check if `cache_dirs` exists in the constructor. 

I don't think that is elegant but keeps compatibility with previous versions while new tests will fail during the development if `cache_dirs` is missing, not being harm in my point of view.

Reply:
Got it. In case `Asset` is just internal API you would be able to break it.